### ISSUE TITLE
A bug that appears when running with CUDA and does not when using cpu

### ISSUE THREAD

**Issue by shiranD** [pytorch/pytorch#5239]
Hi,
When I run a snippet like 

def calc_gradient_penalty(netD, real_data, fake_data):

    wrdseq, batch, embd = real_data.shape
    alpha = torch.rand(wrdseq, batch, embd)
    alpha = alpha.expand(real_data.shape)
    alpha = alpha.cuda(gpu) if use_cuda else alpha
    interpolates = alpha * real_data + ((1 - alpha) * fake_data)

    if use_cuda:
        interpolates = interpolates.cuda(gpu)
    interpolates = autograd.Variable(interpolates, requires_grad=True)

    hiddenD = netD.init_hidden(train_batch_size)
    disc_interpolates, _ = netD(interpolates, hiddenD)

    gradients = autograd.grad(outputs=disc_interpolates, inputs=interpolates,
                              grad_outputs=torch.ones(disc_interpolates.shape).cuda(gpu) if use_cuda else torch.ones(
                                  disc_interpolates.size()),
                              create_graph=True, retain_graph=True, only_inputs=True)[0]

    gradient_penalty = ((gradients.norm(2, dim=1) - 1) ** 2).mean() * LAMBDA
    return gradient_penalty

gradient_penalty = calc_gradient_penalty(netD, _data.data, _fake.data)
gradient_penalty.backward()`

with CUDA I get:
CudnnRNNLegacyBackward is not differentiable twice
but with cpu there is no error.
What should I do to fix it? 
It seems to be related to this cpp script: ./pytorch/torch/csrc/autograd/python_function.cpp
How can I make CudnnRNNLegacyBackward differentiable twice since I assume that the equivalent object under cpu can do that as I get no errors

Would appreciate your help so deeply
Shiran

**Comment by SsnL** [comment]
Well the reason is CPU (and non-cudnn GPU) RNNs are implemented entirely using autograd, all operations of which supports double backward. If you really want to use the non-cudnn GPU RNN, just set `torch.backends.cudnn.enabled=False`. Then you should be able to do the same thing on GPU. That said, TBH, I don't even know if gradient penalty with recurrent discriminator would help.

**Comment by shiranD** [comment]
First,
Thanks for replying and sharing with me what you know – it’s very helpful (I also realized that you are from the developers community – which is awesome!)
I’m not sure I understood the last sentence “I don't even know if gradient penalty with recurrent discriminator would help”. I guess I’m not sure what you meant by that.
Also, if I set torch.backends.cudnn.enabled=False can I still use a gpu? Would that be meaningful? Because if the model is on CPU and the data is on GPU they might be a lot of overhead and at some point both would end up being on the CPU? Does that make sense? I guess I really want to use the gpus I have but if it would increase runtime I’m missing the point and I prefer just using a cpu
I guess I would like to know what is your take on that

Thanks,
Shiran

**Comment by SsnL** [comment]
@shiranD Yes, you can still use a GPU. That flag controls whether some operations are dispatched to our GPU implementation (THNN) or NVIDIA's GPU implementation (CUDNN). Usually the second is faster, that's why we dispatch to CUDNN by default. However, in your case, CUDNN RNN (which is considered as a single autograd op in PyTorch) doesn't support double backward, but the THNN RNN (which is just a series of autograd operations) does. So it works, albeit some potential slow downs.

About the last sentence in my comment above, personally and empirically, I'm not sure how much gradient penalty helps. There are some intuition around it, such as smoothing D around the manifold etc., and some people have been reporting that it reduces sensitivity to hyperparameters a bit in certain z=>img GAN settings. But I haven't found it to very effective in non z=>img setting, such as simply adversarial loss. Considering that sequence data is a completely different domain that image, I'm just not very certain that whether it will help.

**Comment by shiranD** [comment]
Wow! This is really good stuff that I learn from you.
I now know that I will try one with a pure gpu not using the grad_penalty and another run with it (but using the cpu rnn as an object)
However, when you mentioned a simple loss, did you mean just the Wasserstein loss and the original loss which is derived in Goodfellow paper?
I mean, reading the Wass paper really convinced me the Wass loss makes a lot of sense. What is your experience with it?

**Comment by SsnL** [comment]
@shiranD I actually mean the non GAN setting, but with adversarial loss. For example in style transfer, you may want some thing to be indistinguishable with another, but it's not a generative model. Yes, when I read WGAN paper, I was excited as well. The theory is nice and clean and intuitively makes sense. However, you need to keep in mind that very little of GAN theory applies to real settings. For example, many intuition come from that D is the inner optimizing loop (min-max formulation), but this is nearly never what is done in practice. In the end, due to the lack of a nice and complete theory, the theories and intuitions tell us what we should try and what might work. But we still need to try to see whether it works.

**Comment by shiranD** [comment]
Right that’s what many people say, but I’m still excited to start and set my foot in and see how it goes.
Thanks for your good advice and for sharing with me your experience.
I might ask more questions in the future – and I hope it’s alright.
Best,
Shiran

**Comment by soumith** [comment]
> I might ask more questions in the future – and I hope it’s alright.

Sure thing :), if you have questions, use https://discuss.pytorch.org in the future.

**Comment by SsnL** [comment]
@shiranD You are welcome. If you have further questions, we can discuss via forum messages and / or email.
